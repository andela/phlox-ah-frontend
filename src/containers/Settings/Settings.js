import { connect } from 'react-redux';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Switch from 'react-switch';
import './Settings.scss';
import { getUserNotificationStatus, optInForNotification, optOutFromNotification } from '../../requests/NotificationRequests';

/**
 * @class Settings
 * @extends {Component}
 */
class Settings extends Component {
  /**
   * @description - This method runs first in the class
   * @returns {object} Settings
   * @memberof Settings
   */
  constructor() {
    super();

    this.state = { checked: false };
    this.toggleButton = this.toggleButton.bind(this);
  }

  /**
   * @description - This method get runs before the component mount
   * @returns {object} void
   * @memberof Settings
   */
  componentDidMount() {
    this.props.getUserNotificationStatus(this.props.user.username);
  }

  /**
   * @description - This method sets the input values
   * @param {boolean} checked
   * @returns {object} void
   * @memberof Settings
   */
  toggleButton(checked) {
    this.setState({ checked });
    if (checked) {
      this.props.optInForNotification();
    } else {
      this.props.optOutFromNotification();
    }
  }

  /**
  * @description - This method runs whenever the redux state changes
  * @returns {object} state
  * @param {object} props
  * @param {object} state
  */
  static getDerivedStateFromProps(props, state) {
    if (props.notification.success) {
      return { checked: props.notification.emailNotification };
    }
    return state;
  }

  /**
   * @description - This method renders the jsx for this component
   * @returns {jsx} - jsx
   * @memberof Settings
   */
  render() {
    return (
      <div className="settings">
        <header className="title">
          <h2>Settings</h2>
        </header>
        <div className="notification">
          <h4 className="subtitle">Turn On All Notifications</h4>
          <div className="notification-content">
            <p className="description">Weâ€™ll email you when there are notifications
                on your stories and publications, and when you have new followers.</p>
            <div className="button">
              <Switch
                onChange={this.toggleButton}
                checked={this.state.checked}
                onColor="#3B8CA1"
                height={30}
              />
            </div>
          </div>
          <hr />
        </div>
      </div>
    );
  }
}

Settings.propTypes = {
  getUserNotificationStatus: PropTypes.func.isRequired,
  notification: PropTypes.object,
  optInForNotification: PropTypes.func,
  optOutFromNotification: PropTypes.func,
  user: PropTypes.object
};

const mapStateToProps = state => ({
  user: state.user,
  notification: state.notification
});

export default connect(mapStateToProps, {
  getUserNotificationStatus,
  optInForNotification,
  optOutFromNotification
})(Settings);
