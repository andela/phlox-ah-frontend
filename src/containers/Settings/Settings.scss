div.settings {
  font-family: Fira Mono;
  margin: 0px auto;
  margin: 50px;
  background: #ffffff;
  padding: 130px;
  padding-top: 25px;
  .title {
    margin-bottom: 30px;
    letter-spacing: 2px;
    text-align: center;
  }
  .notification {
    display: flex;
    flex-direction: column;
    margin-bottom: 20px;
  }
  .notification-content {
    display: flex;
    align-items: center;
  }
  .notification-content .button {
    flex-grow: 1;
    text-align: right;
  }
  .subtitle {
    margin-bottom: 20px;
    letter-spacing: 1px;
  }
  hr {
    margin-top: 20px;
    border: 0.5px solid #CCCCCC;
  }
  .social-link {
    margin-top: 25px;
  }
}
