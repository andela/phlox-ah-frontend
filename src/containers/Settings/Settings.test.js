import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import Settings from './Settings';

const mockStore = configureMockStore();
const store = mockStore({
  notification: {
    error: null,
    message: 'successfully opted out of notification',
    success: true,
    emailNotification: false
  },
  user: {
    isAuth: true,
    loading: false,
    id: 6,
    email: 'admin@authorshaven.com',
    username: 'admin1',
    role: 'Admin'
  }
});
let component;
let myComponent;

beforeEach(() => {
  component = shallow(
    <Provider store={store}>
      <Settings />
    </Provider>
  );
  myComponent = component.dive({ context: { store } }).dive();
});

it('should render without throwing an error', () => {
  expect(component).toMatchSnapshot();
});
