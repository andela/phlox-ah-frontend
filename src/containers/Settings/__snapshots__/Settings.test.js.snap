// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without throwing an error 1`] = `
<div
  className="settings"
>
  <header
    className="title"
  >
    <h2>
      Settings
    </h2>
  </header>
  <div
    className="notification"
  >
    <h4
      className="subtitle"
    >
      Turn On All Notifications
    </h4>
    <div
      className="notification-content"
    >
      <p
        className="description"
      >
        Weâ€™ll email you when there are notifications on your stories and publications, and when you have new followers.
      </p>
      <div
        className="button"
      >
        <ReactSwitch
          activeBoxShadow="0 0 2px 3px #3bf"
          boxShadow={null}
          checked={false}
          checkedIcon={
            <svg
              height="100%"
              style={
                Object {
                  "position": "absolute",
                  "top": 0,
                }
              }
              viewBox="-2 -5 17 21"
              width="100%"
            >
              <path
                d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
                fill="#fff"
                fillRule="evenodd"
              />
            </svg>
          }
          disabled={false}
          height={30}
          id="notificationStatus"
          offColor="#888"
          offHandleColor="#fff"
          onChange={[Function]}
          onColor="#3B8CA1"
          onHandleColor="#fff"
          uncheckedIcon={
            <svg
              height="100%"
              style={
                Object {
                  "position": "absolute",
                  "top": 0,
                }
              }
              viewBox="-2 -5 14 20"
              width="100%"
            >
              <path
                d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
                fill="#fff"
                fillRule="evenodd"
              />
            </svg>
          }
          width={56}
        />
      </div>
    </div>
    <hr />
  </div>
</div>
`;
